### PR TITLE
feat: Ec 2004 crack control

### PR DESCRIPTION
**Pull request by DanielGMorenaFhecor**
I have started working on some functions from EC2-1992-1-1:2004 Section 7 - SLS. 7.3 Crack Control.

Also, testing functions have been added to validate expected input-outputs, and ValueErrors are raised for non-correct inputs.

In addition, detailed Google-format Docstrings have been added.

### CODE REVIEW

**Comment by DanielGMorenaFhecor** [comment]
Sorry, @mortenengen! I just requested you a code revision by mistake! 
I tried to add @aperezcaldentey and @MestreCarlos to review my code using this same PR, but I think they have not been added as Collaborators.

Would it be possible to add them?

Thank you in advance!

**Comment by mortenengen** [comment]
Of course 😃

**Review by MestreCarlos — APPROVED** [review]
I did a cross-validation of the code by applying it to the calculation of the crack width in a rectangular concrete beam subjected to bending

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
     "python.linting.pylintEnabled": true,
-    "python.linting.flake8Enabled": true
+    "python.linting.flake8Enabled": true,
+    "[python]": {
+        "editor.formatOnSave": true,
+        "editor.codeActionsOnSave": {
+            "source.organizeImports": true
+        },
+        "editor.defaultFormatter": "ms-python.python",
+    },
```
Comment (mortenengen) [inline .vscode/settings.json:16]: Please remove the settings related to autoformating. We are using Black, and this can be configured to reformat the code when it is run.

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+numpy==1.23.5
+scipy==1.9.3
```
Comment (mortenengen) [inline requirements.txt:2]: Please use a `>=` instead of `==` to avoid dependency conflicts with other packages. Also, make sure that you actually pin the lowest relevant version, i.e. the lowest version that is compatible with the functionality that we depend on.

File: structuralcodes/material/concrete/__init__.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+"""Concrete material"""
```
Comment (mortenengen) [inline structuralcodes/material/concrete/__init__.py:1]: Note that this file is now in structuralcodes/material/concrete. Please make sure that you rename your folder accordingly, and the diff disappears.

File: structuralcodes/material/concrete/_concrete.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+"""Core implementation of the concrete material"""
```
Comment (mortenengen) [inline structuralcodes/material/concrete/_concrete.py:1]: Same comment as for concrete/__init__.py

File: structuralcodes/material/concrete/_concreteMC2010.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
+"""The concrete class for Model Code 2020 Concrete Material"""
```
Comment (mortenengen) [inline structuralcodes/material/concrete/_concreteMC2010.py:1]: Same comment as for concrete/__init__.py

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 6]
```diff
@@ -0,0 +1,933 @@
...
+Chapter 7.3 - Crack control"""
+import math
+import typing as t
+
+import numpy as np
+import scipy.interpolate
+
+
+def w_max(exposure_class: str, load_combination: str) -> float:
+    """Computes the recomended value of the maximum crack width.
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:11]: Typo: recomended -> recommended

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 7]
```diff
@@ -0,0 +1,933 @@
...
+    A_ct: float, sigma_s: float, fct_eff: float, _k: float, kc: float
+) -> float:
+    """Computes the minimum area of reinforcing steel within the tensile zone
+    for control of cracking areas
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.1)
+
+    Args:
+        A_ct (float): is the area of concrete within the tensile zone in mm2.
+            The tensile zone is that parg of the section which is calculated
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:71]: Typo: parg -> part

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 8]
```diff
@@ -0,0 +1,933 @@
...
+    Raises:
+        ValueError: if h is less than 0
+    """
+    if h < 0:
+        raise ValueError(f'h={h} cannot be less than 0mm')
+    if h <= 300:
+        return 1
+    if h < 800:
+        interpol = scipy.interpolate.interp1d((300, 800), (1, 0.65))
+        return (float)(interpol(h))
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:140]: Consider removing the parentheses around `float`, and consider removing also `float` since `interpol` here most likely returns a `float`.

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+    return 0.65
+
+
+def kc_tension() -> float:
+    """Computes the coefficient which takes account of the stress
+    distribution within the section immediately prior to cracking and
+    the change of the lever arm in pure dtension.
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:147]: Typo: dtension -> tension

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 10]
```diff
@@ -0,0 +1,933 @@
...
+    combination of loads of imposed deformations and taking into account the
+    effects of tension stiffening. Only the additional tensile strain beyond
+    the state of zero strain of the concrete is considered. ecm is the mean
+    strain in the concrete between the cracks.
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.9)
+
+    Args:
+        sigma_s (float): is the stress in MPa in the tension reinforcement
+            assuming a cracked section. FOr pretensioned members, s_steel may
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:614]: Typo: FOr -> For

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 11]
```diff
@@ -0,0 +1,933 @@
...
+            be replaced by increment of s_steel stress variation in
+            prestressing tendons from the state of zero strain of the
+            concrete at the same level.
+        _alpha_e (float): is the ratio Es/Ecm
+        _rho_p_eff (float): effective bond ratio between areas given by the
+            Eq. (7.10)
+        _kt (float): is a factor dependent on the load duration
+        fct_eff (float): is the mean value of the tensile strength in MPa
+            of the concrete effectvie at the time when the cracks may
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:623]: Typo: effectvie -> effective

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 12]
```diff
@@ -0,0 +1,933 @@
...
+            first be expected to occur: fct_eff=fctm or fctm(t) if
+            crack is expected earlier than 28 days.
+        Es: steel elastic mudulus in MPa
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:626]: Typo: mudulus -> modulus

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 13]
```diff
@@ -0,0 +1,933 @@
...
+def w_spacing(c: float, phi: float) -> float:
+    """Computes the distance threshold from which the
+    maximum crack spacing is constant.
+
+    EUROCODE 2 1992-1-1:2004, Sect. (7.3.4-3)
+
+    Args:
+        c (float): cover of  the longitudinal reinforcement in mm
+        phi (float): is the bar diameter in mm. Where mixed bar diameters
+            used, then it should be replaced for an equivalente bar diameter.
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:667]: Typo: equivalente -> equivalent

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 14]
```diff
@@ -0,0 +1,933 @@
...
+    """Get the k1 coefficient which takes account of the bond properties
+    of the bounded reinforcement
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.11-k1)
+
+    Args:
+        bond_type (str): the bond property of the reinforcement.
+        Possible values:
+            - 'bond': for high bond bars
+            - 'plane': for bars with an effectively plain surface (e.g.
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:731]: Keyword should be `plain` instead of `plane`.

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 15]
```diff
@@ -0,0 +1,933 @@
...
+def k2(epsilon_r: float) -> float:
+    """Computes a coefficient which takes into account of the
+    distribution of strain:
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.13)
+
+    Args:
+        epsilon_r (float): ratio epsilon_2/epsilon_1 where epsilon_1 is
+            thre greater and epsilon_2 is the lesser strain at the boundaries
+            of the section considererd, assessed on the basis of a cracked
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:762]: Typo: considererd -> considered

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 16]
```diff
@@ -0,0 +1,933 @@
...
+
+
+def sr_max_close(
+    c: float,
+    phi: float,
+    _rho_p_eff: float,
+    _k1: float,
+    _k2: float,
+    _k3: float,
+    _k4: float,
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:802]: Consider making the `k`s optional, e.g. `_k1: t.Optional[float] = None`, and use the relevant `k` function for computing it's value if no value is provided.

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 17]
```diff
@@ -0,0 +1,933 @@
...
+    if _k2 < 0.5 or _k2 > 1:
+        raise ValueError(f'_k2={_k2} is not between 0.5 and 1.0')
+
+    return _k3 * c + _k1 * _k2 * _k4 * phi / _rho_p_eff
+
+
+def sr_max_far(h: float, x: float) -> float:
+    """Computes the maximum crack spacing in cases where bonded reinforcement
+    is fixed at reasonably close centres within the tension zone
+    (w_spacing>5(c+phi/2)).
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:853]: The docstring seems to be a duplicate of `sr_max_close`, please revise.

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 18]
```diff
@@ -0,0 +1,933 @@
...
+            reinforcement, fyk. A lower value may, however, be needed to
+            satisfy the crack width limits according to the maximum
+            bar size of spacing (see 7.3.3 (2)).
+        fct_eff (float): is the mean value of the tensile strength in MPa of
+            the concrete effective at the time when the cracks may first be
+            expected to occur: fct,eff=fct or lower (fct(t)), is cracking
+            is expected earlier than 28 days.
+        _k (float): is the coefficient which allow for the effect of
+            non-uniform self-equilibrating stresses, which lead to a
+            reduction of restraint forces. Use 'k_crack_min_steel_area'
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:85]: I cannot find a definition of `k_crack_min_steel_area`.

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 19]
```diff
@@ -0,0 +1,933 @@
...
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:326]: I cannot find a definition of `k_crack_min_steel_area`

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py [hunk 20]
```diff
@@ -0,0 +1,933 @@
...
+def As_min_p(
+    A_ct: float,
+    sigma_s: float,
+    fct_eff: float,
+    _k: float,
+    kc: float,
+    Ap: float,
+    phi_s: float,
+    phi_p: float,
+    xi: float,
```
Comment (mortenengen) [inline structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:302]: Mismatch between the function signature and the docstring. Please revise.

File: .vscode/settings.json (hunk 1 above)
Comment (DanielGMorenaFhecor) [reply .vscode/settings.json:16]: I have just modified the file so I keep the original config! 

I think we already discussed it in the previous meeting, but I am not able to format files in VSCode with black with the original config (not even using the command _make _form__ or any other alternative). 

Maybe I am just wrong, but if you do not mind, we can talk about it tomorrow.

File: requirements.txt (hunk 2 above)
Comment (DanielGMorenaFhecor) [reply requirements.txt:2]: Ok! Now the values are:

`numpy>=1.20.0
scipy>=1.6.0`

Those point to the first versions of the packages that support our older python version that is being tested, which I think is 3.7 . 

If we should use older packages versions please, tell me and I will change it! 😃

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py (hunk 16 above)
Comment (DanielGMorenaFhecor) [reply structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py:802]: Great idea! 

What do you think about adding the option of being None only to `_k3` and `_k4`? Parameters `_k1` and `_k2` require as input values `bond_type` and `epsilon_r` respectively to be computed.

I have also included the tests if these values are None.

**Comment by DanielGMorenaFhecor** [comment]
I have just updated the greek chars conventions to the ones Alejandro sent. It mainly involved the EUROCODE 2 1992-1-1:2004, 7.9 Equation and its references: `` esm_ecm -> eps_sm_eps_cm ``

File: structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py (hunk 16 above)
Comment (mortenengen) [reply structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py]: Good idea, keep `_k3` and `_k4` optional and the rest required 😃

**Review by mortenengen — APPROVED** [review]
